libpam_internal/pam_debug.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * This provides the necessary functions to do debugging in PAM.
+ * Cristian Gafton <gafton@redhat.com>
+ */
+
+#include "config.h"
+
+#ifdef PAM_DEBUG
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <stdarg.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include "security/_pam_types.h"
+
+/*
+ * This is for debugging purposes ONLY. DO NOT use on live systems !!!
+ * You have been warned :-) - CG
+ *
+ * to get automated debugging to the log file, it must be created manually.
+ * _PAM_LOGFILE must exist and be writable to the programs you debug.
+ */
+
+#ifndef _PAM_LOGFILE
+#define _PAM_LOGFILE "/var/run/pam-debug.log"
+#endif
+
+UNUSED
+void _pam_output_debug_info(const char *file, const char *fn, const int line);
+UNUSED
+PAM_FORMAT((printf, 1, 2))
+void _pam_output_debug(const char *format, ...);
+
+UNUSED
+void _pam_output_debug_info(const char *file, const char *fn
+			    , const int line)
+{
+    FILE *logfile;
+    int must_close = 1, fd;
+
+#ifdef O_NOFOLLOW
+    if ((fd = open(_PAM_LOGFILE, O_WRONLY|O_NOFOLLOW|O_APPEND)) != -1) {
+#else
+    if ((fd = open(_PAM_LOGFILE, O_WRONLY|O_APPEND)) != -1) {
+#endif
+	if (!(logfile = fdopen(fd,"a"))) {
+	    logfile = stderr;
+	    must_close = 0;
+	    close(fd);
+	}
+    } else {
+        logfile = stderr;
+	must_close = 0;
+    }
+    fprintf(logfile,"[%s:%s(%d)] ",file, fn, line);
+    fflush(logfile);
+    if (must_close)
+        fclose(logfile);
+}
+
+UNUSED
+PAM_FORMAT((printf, 1, 2))
+void _pam_output_debug(const char *format, ...)
```

Comment:
Function attributes likes `UNUSED` and `PAM_FORMAT` are needed for function prototypes only.

Reply:
Removed